### PR TITLE
Add substrate connect support to staking dashboard

### LINKED ISSUE #149: Integrate Substrate Connect as connection option
https://github.com/paritytech/substrate-connect/blob/main/projects/burnr/src/hooks/api/useApiCreate.ts
improve security through using lightclient instead of rpc for dashboard.

### PR DESCRIPTION
**Pull request by wirednkod**
Add light client (substrate connect) support to available chains (Polkadot, Kusama, Westend);
Fixes #149;

Some screenshots that show the design result as well:

Screenshot 2022-07-10 at 3 06 59 AM

Screenshot 2022-07-10 at 3 07 39 AM

### CODE REVIEW

**Comment by wirednkod** [comment]
I can see that there are some numbers (e.g. Minimum Active Bond, Nominations (active, Inactive, list), Active Validators number) that is not returned to the dashboard when light client is used.
Im not sure if thats a light client's issue or the way PolkadotJS is implemented (and we use it with ScProbiver).
I'll add a reminder to myself for that

**Comment by rossbulat** [comment]
> I can see that there are some numbers (e.g. Minimum Active Bond, Nominations (active, Inactive, list), Active Validators number) that is not returned to the dashboard when light client is used. Im not sure if thats a light client's issue or the way PolkadotJS is implemented (and we use it with ScProbiver).
> 
> I'll add a reminder to myself for that - and we should probably put some "experimental" asterisk when light client is used (as we did with polkadotJS/apps)

These stats just fetch everything from `erasStakers` and then use a web worker to loop through the exposures and accumulate the stats. 

The fetching is done [here](https://github.com/paritytech/polkadot-staking-dashboard/blob/828ba86bc2d25eba9455c0ba8b03ddd2f718f0fd/src/contexts/Staking/index.tsx#L259), and the callback defined [here](https://github.com/paritytech/polkadot-staking-dashboard/blob/828ba86bc2d25eba9455c0ba8b03ddd2f718f0fd/src/contexts/Staking/index.tsx#L166).

Is the light client fetching erasStakers successfully?
```
const _exposures = await api.query.staking.erasStakers.entries(
      metrics.activeEra.index
    );
```

**Comment by wirednkod** [comment]
> > I can see that there are some numbers (e.g. Minimum Active Bond, Nominations (active, Inactive, list), Active Validators number) that is not returned to the dashboard when light client is used. Im not sure if thats a light client's issue or the way PolkadotJS is implemented (and we use it with ScProbiver).
> > I'll add a reminder to myself for that - and we should probably put some "experimental" asterisk when light client is used (as we did with polkadotJS/apps)
> 
> These stats are derived from `erasStakers`, which is passed to a web worker to loop through the exposures and accumulate the stats.
> 
> The fetching is done [here](https://github.com/paritytech/polkadot-staking-dashboard/blob/828ba86bc2d25eba9455c0ba8b03ddd2f718f0fd/src/contexts/Staking/index.tsx#L259), and the callback defined [here](https://github.com/paritytech/polkadot-staking-dashboard/blob/828ba86bc2d25eba9455c0ba8b03ddd2f718f0fd/src/contexts/Staking/index.tsx#L166).
> 
> Is the light client fetching erasStakers successfully?
> 
> ```
> const _exposures = await api.query.staking.erasStakers.entries(
>    metrics.activeEra.index
> );
> ```

[This](https://github.com/paritytech/smoldot/pull/2491) fixes the issue. Next version of substrate-connect that will be released today will have that in.

**Comment by wirednkod** [comment]
Added latest version of `rpc-provider` and stylised accordingly the light client based on the left menubar.
Indication of "light" next to sidebar logo when light client is active have been added.
In addition `localStorage` has a new item called `isLightClient` in order to keep the latest choice of user.

**Comment by rossbulat** [comment]
Hi Nikos, I am having issues running this in Brave browser. The syncing hangs with 0 peers. Do we know what is causing this? On Chrome it is working well. Screenshots for context:

light1

Screenshot 2022-08-09 at 14 33 12

Latest logs:
```
...
14:33:59 427 Debug - json-rpc-polkadot JSON-RPC <= {"jsonrpc":"2.0","id":"health-check:96422","result":{"isSyncing":false,"peers":0,"shouldHavePeers":t…
14:33:59 430 Debug - json-rpc-ksmcc3 JSON-RPC <= {"jsonrpc":"2.0","id":"health-check:96419","result":{"isSyncing":false,"peers":0,"shouldHavePeers":t…
14:33:59 430 Debug - json-rpc-rococo_v2_2 JSON-RPC <= {"jsonrpc":"2.0","id":"health-check:96420","result":{"isSyncing":false,"peers":0,"shouldHavePeers":t…
14:33:59 431 Debug - json-rpc-westend2 JSON-RPC <= {"jsonrpc":"2.0","id":"health-check:96421","result":{"isSyncing":false,"peers":0,"shouldHavePeers":t…
```

**Comment by rossbulat** [comment]
Perhaps a separate issue, I am also getting the following console errors on `yarn start`:

```
WARNING in ./node_modules/@substrate/connect-extension-protocol/dist/mjs/index.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect-extension-protocol/src/index.ts' file: Error: ENOENT: no such file or directory, open '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect-extension-protocol/src/index.ts'
 @ ./node_modules/@substrate/connect/dist/mjs/connector/index.js 3:0-71 22:141-155
 @ ./node_modules/@substrate/connect/dist/mjs/index.js 86:0-37 86:0-37
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/ScProvider.js 3:0-68 14:46-60 16:0-26 67:78-92
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/index.js 3:0-32 3:0-32
 @ ./node_modules/@polkadot/rpc-provider/bundle.js 6:0-58 6:0-58
 @ ./node_modules/@polkadot/rpc-provider/index.js 4:0-28 4:0-28
 @ ./node_modules/@polkadot/api/bundle.js 5:0-66 5:0-66 5:0-66
 @ ./node_modules/@polkadot/api/index.js 4:0-28 4:0-28
 @ ./src/contexts/Api/index.tsx 10:0-55 70:21-31 112:26-36
 @ ./src/Providers.tsx 14:0-53 57:6-12 77:20-26 81:39-50
 @ ./src/App.tsx 8:0-38 21:35-44
 @ ./src/index.tsx 8:0-22 16:34-37

WARNING in ./node_modules/@substrate/connect/dist/mjs/WellKnownChain.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/WellKnownChain.ts' file: Error: ENOENT: no such file or directory, open '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/WellKnownChain.ts'
 @ ./node_modules/@substrate/connect/dist/mjs/index.js 85:0-53 85:0-53
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/ScProvider.js 3:0-68 14:46-60 16:0-26 67:78-92
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/index.js 3:0-32 3:0-32
 @ ./node_modules/@polkadot/rpc-provider/bundle.js 6:0-58 6:0-58
 @ ./node_modules/@polkadot/rpc-provider/index.js 4:0-28 4:0-28
 @ ./node_modules/@polkadot/api/bundle.js 5:0-66 5:0-66 5:0-66
 @ ./node_modules/@polkadot/api/index.js 4:0-28 4:0-28
 @ ./src/contexts/Api/index.tsx 10:0-55 70:21-31 112:26-36
 @ ./src/Providers.tsx 14:0-53 57:6-12 77:20-26 81:39-50
 @ ./src/App.tsx 8:0-38 21:35-44
 @ ./src/index.tsx 8:0-22 16:34-37

WARNING in ./node_modules/@substrate/connect/dist/mjs/connector/extension.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/extension.ts' file: Error: ENOENT: no such file or directory, open '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/extension.ts'
 @ ./node_modules/@substrate/connect/dist/mjs/connector/index.js 2:0-69 31:51-68
 @ ./node_modules/@substrate/connect/dist/mjs/index.js 86:0-37 86:0-37
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/ScProvider.js 3:0-68 14:46-60 16:0-26 67:78-92
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/index.js 3:0-32 3:0-32
 @ ./node_modules/@polkadot/rpc-provider/bundle.js 6:0-58 6:0-58
 @ ./node_modules/@polkadot/rpc-provider/index.js 4:0-28 4:0-28
 @ ./node_modules/@polkadot/api/bundle.js 5:0-66 5:0-66 5:0-66
 @ ./node_modules/@polkadot/api/index.js 4:0-28 4:0-28
 @ ./src/contexts/Api/index.tsx 10:0-55 70:21-31 112:26-36
 @ ./src/Providers.tsx 14:0-53 57:6-12 77:20-26 81:39-50
 @ ./src/App.tsx 8:0-38 21:35-44
 @ ./src/index.tsx 8:0-22 16:34-37

WARNING in ./node_modules/@substrate/connect/dist/mjs/connector/index.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/index.ts' file: Error: ENOENT: no such file or directory, open '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/index.ts'
 @ ./node_modules/@substrate/connect/dist/mjs/index.js 86:0-37 86:0-37
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/ScProvider.js 3:0-68 14:46-60 16:0-26 67:78-92
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/index.js 3:0-32 3:0-32
 @ ./node_modules/@polkadot/rpc-provider/bundle.js 6:0-58 6:0-58
 @ ./node_modules/@polkadot/rpc-provider/index.js 4:0-28 4:0-28
 @ ./node_modules/@polkadot/api/bundle.js 5:0-66 5:0-66 5:0-66
 @ ./node_modules/@polkadot/api/index.js 4:0-28 4:0-28
 @ ./src/contexts/Api/index.tsx 10:0-55 70:21-31 112:26-36
 @ ./src/Providers.tsx 14:0-53 57:6-12 77:20-26 81:39-50
 @ ./src/App.tsx 8:0-38 21:35-44
 @ ./src/index.tsx 8:0-22 16:34-37

WARNING in ./node_modules/@substrate/connect/dist/mjs/connector/smoldot-light.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/smoldot-light.ts' file: Error: ENOENT: no such file or directory, open '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/smoldot-light.ts'
 @ ./node_modules/@substrate/connect/dist/mjs/connector/index.js 1:0-71 32:9-24
 @ ./node_modules/@substrate/connect/dist/mjs/index.js 86:0-37 86:0-37
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/ScProvider.js 3:0-68 14:46-60 16:0-26 67:78-92
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/index.js 3:0-32 3:0-32
 @ ./node_modules/@polkadot/rpc-provider/bundle.js 6:0-58 6:0-58
 @ ./node_modules/@polkadot/rpc-provider/index.js 4:0-28 4:0-28
 @ ./node_modules/@polkadot/api/bundle.js 5:0-66 5:0-66 5:0-66
 @ ./node_modules/@polkadot/api/index.js 4:0-28 4:0-28
 @ ./src/contexts/Api/index.tsx 10:0-55 70:21-31 112:26-36
 @ ./src/Providers.tsx 14:0-53 57:6-12 77:20-26 81:39-50
 @ ./src/App.tsx 8:0-38 21:35-44
 @ ./src/index.tsx 8:0-22 16:34-37

WARNING in ./node_modules/@substrate/connect/dist/mjs/connector/specs/index.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/specs/index.ts' file: Error: ENOENT: no such file or directory, open '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/specs/index.ts'
 @ ./node_modules/@substrate/connect/dist/mjs/connector/smoldot-light.js 1:0-43 146:25-32
 @ ./node_modules/@substrate/connect/dist/mjs/connector/index.js 1:0-71 32:9-24
 @ ./node_modules/@substrate/connect/dist/mjs/index.js 86:0-37 86:0-37
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/ScProvider.js 3:0-68 14:46-60 16:0-26 67:78-92
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/index.js 3:0-32 3:0-32
 @ ./node_modules/@polkadot/rpc-provider/bundle.js 6:0-58 6:0-58
 @ ./node_modules/@polkadot/rpc-provider/index.js 4:0-28 4:0-28
 @ ./node_modules/@polkadot/api/bundle.js 5:0-66 5:0-66 5:0-66
 @ ./node_modules/@polkadot/api/index.js 4:0-28 4:0-28
 @ ./src/contexts/Api/index.tsx 10:0-55 70:21-31 112:26-36
 @ ./src/Providers.tsx 14:0-53 57:6-12 77:20-26 81:39-50
 @ ./src/App.tsx 8:0-38 21:35-44
 @ ./src/index.tsx 8:0-22 16:34-37

WARNING in ./node_modules/@substrate/connect/dist/mjs/connector/types.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/types.ts' file: Error: ENOENT: no such file or directory, open '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/connector/types.ts'
 @ ./node_modules/@substrate/connect/dist/mjs/connector/index.js 4:0-27 4:0-27
 @ ./node_modules/@substrate/connect/dist/mjs/index.js 86:0-37 86:0-37
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/ScProvider.js 3:0-68 14:46-60 16:0-26 67:78-92
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/index.js 3:0-32 3:0-32
 @ ./node_modules/@polkadot/rpc-provider/bundle.js 6:0-58 6:0-58
 @ ./node_modules/@polkadot/rpc-provider/index.js 4:0-28 4:0-28
 @ ./node_modules/@polkadot/api/bundle.js 5:0-66 5:0-66 5:0-66
 @ ./node_modules/@polkadot/api/index.js 4:0-28 4:0-28
 @ ./src/contexts/Api/index.tsx 10:0-55 70:21-31 112:26-36
 @ ./src/Providers.tsx 14:0-53 57:6-12 77:20-26 81:39-50
 @ ./src/App.tsx 8:0-38 21:35-44
 @ ./src/index.tsx 8:0-22 16:34-37

WARNING in ./node_modules/@substrate/connect/dist/mjs/index.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/index.ts' file: Error: ENOENT: no such file or directory, open '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect/src/index.ts'
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/ScProvider.js 3:0-68 14:46-60 16:0-26 67:78-92
 @ ./node_modules/@polkadot/rpc-provider/substrate-connect/index.js 3:0-32 3:0-32
 @ ./node_modules/@polkadot/rpc-provider/bundle.js 6:0-58 6:0-58
 @ ./node_modules/@polkadot/rpc-provider/index.js 4:0-28 4:0-28
 @ ./node_modules/@polkadot/api/bundle.js 5:0-66 5:0-66 5:0-66
 @ ./node_modules/@polkadot/api/index.js 4:0-28 4:0-28
 @ ./src/contexts/Api/index.tsx 10:0-55 70:21-31 112:26-36
 @ ./src/Providers.tsx 14:0-53 57:6-12 77:20-26 81:39-50
 @ ./src/App.tsx 8:0-38 21:35-44
 @ ./src/index.tsx 8:0-22 16:34-37
```

**Comment by wirednkod** [comment]
> @wirednkod Perhaps a separate issue, I am also getting the following console errors on `yarn start`:
> 
> ```
> WARNING in ./node_modules/@substrate/connect-extension-protocol/dist/mjs/index.js
> Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
> Failed to parse source map from '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect-extension-protocol/src/index.ts' file: Error: ENOENT: no such file or directory, open '/Users/macbookpro/polkadot/polkadot-staking-dashboard/node_modules/@substrate/connect-extension-protocol/src/index.ts'

We should open an issue for this.. Im not sure what is really triggering it... We have made some adjustments that may help warnings disappear in next version

**Comment by wirednkod** [comment]
> Hi Nikos, I am having issues running this in Brave browser. The syncing hangs with 0 peers. Do we know what is causing this? On Chrome it is working well.

There is some peculiar decision making with [Brave since 1.36](https://github.com/brave/brave-browser/issues/19990). In substrate connect extension we [prepare](https://github.com/paritytech/substrate-connect/pull/1180) a similar "note" on that for Brave users, and we have added also a note on the [landing page](https://paritytech.github.io/substrate-connect/#troubleshooting). 

TLDR; `As of Brave v1.36, extensions and web pages are limited to a maximum of 10 active WebSocket connections`

**Comment by rossbulat** [comment]
> > Hi Nikos, I am having issues running this in Brave browser. The syncing hangs with 0 peers. Do we know what is causing this? On Chrome it is working well.
> 
> There is some peculiar decision making with [Brave since 1.36](https://github.com/brave/brave-browser/issues/19990). In substrate connect extension we [prepare](https://github.com/paritytech/substrate-connect/pull/1180) a similar "note" on that for Brave users, and we have added also a note on the [landing page](https://paritytech.github.io/substrate-connect/#troubleshooting).
> 
> TLDR; `As of Brave v1.36, extensions and web pages are limited to a maximum of 10 active WebSocket connections`

Got it, thank you for the context. Would a good solution be some browser detection in-app and disable light client support where browsers are being problematic?

**Comment by wirednkod** [comment]
Well we can do that, and its an easy solution, but its a pity to lose Brave users, due to a setting. IMO its better to throw a note for Brave users, that there is a setting that may create issues with websockets, rather than disable light client support.  In addition, keep in mind that this is a temp solution, as we have in our roadmap the integration with WebRTC that will overcome these kind of issues

File: src/types/index.ts [hunk 1]
```diff
@@ -24,6 +24,10 @@ export interface Networks {
 
 export interface Network {
   name: string;
+  endpoints: {
+    rpc: string;
+    lightClient: WellKnownChain;
```
Comment (wirednkod) [inline src/types/index.ts:29]: You should now always assume that a light client will exist. 
This should be:
```
    lightClient?: WellKnownChain;
```

File: src/contexts/Api/index.tsx [hunk 2]
```diff
@@ -154,12 +154,14 @@ export const APIProvider = ({ children }: { children: React.ReactNode }) => {
   // connect function sets provider and updates active network.
   const connect = async (_network: NetworkName, _isLightClient?: boolean) => {
     const nodeEndpoint: Network = NETWORKS[_network];
+    const { endpoints } = nodeEndpoint;
+
     let _provider: WsProvider | ScProvider;
-    if (_isLightClient && nodeEndpoint.lightClientEndpoint) {
-      _provider = new ScProvider(nodeEndpoint.lightClientEndpoint);
+    if (_isLightClient && endpoints.lightClient) {
```
Comment (wirednkod) [inline src/contexts/Api/index.tsx:160]: Same for the optional `lightClient` parameter
```
    if (_isLightClient && endpoints?.lightClient) {
```

File: src/types/index.ts (hunk 1 above)
Comment (rossbulat) [reply src/types/index.ts:30]: Nope, we are assuming they exist as all supported chains have a light client

Comment (rossbulat) [reply src/types/index.ts:30]: Is there a case where they would not exist?

**Comment by rossbulat** [comment]
Ready to merge. Please let me know if the optional light client flag is totally necessary, It is probably wise not to add complexity if not needed.

File: src/types/index.ts (hunk 1 above)
Comment (wirednkod) [reply src/types/index.ts:30]: > Is there a case where they would not exist?

The only case is either a custom chain that supports staking or in case we do not want to include light client for a known chain (e.g. polkadot) inside the dashboard. This is of course something that can be altered on a later PR (current PR  implementation supports the network UI when light client does not exist [e.g. shows full width button instead of split to 50/50 of RPC/Light client)

**Comment by jonathanpdunne** [comment]
fwiw Talisman also has a similar flag in our wallet. If there's a way to mirror copy and present a consistent message to users of both products, I'd be open to changing what we have in the wallet.

image

**Comment by wirednkod** [comment]
Same for substrate connect extension. I will open a discussion about that so we can track it better. Dont know though where would be the best place to do so
